src/tools/immutable_object_cache/SimplePolicy.cc

```diff
@@ -0,0 +1,219 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "common/debug.h"
+#include "SimplePolicy.h"
+
+#define dout_context g_ceph_context
+#define dout_subsys ceph_subsys_immutable_obj_cache
+#undef dout_prefix
+#define dout_prefix *_dout << "ceph::cache::SimplePolicy: " << this << " " \
+                           << __func__ << ": "
+
+namespace ceph {
+namespace immutable_obj_cache {
+
+SimplePolicy::SimplePolicy(CephContext *cct, uint64_t block_num,
+                           float watermark)
+  : cct(cct), m_watermark(watermark), m_entry_count(block_num),
+    m_cache_map_lock("rbd::cache::SimplePolicy::m_cache_map_lock"),
+    m_free_list_lock("rbd::cache::SimplePolicy::m_free_list_lock") {
+  ldout(cct, 20) << dendl;
+  m_max_inflight_ops = cct->_conf.get_val<uint64_t>("immutable_object_cache_max_inflight_ops");
+  for(uint64_t i = 0; i < m_entry_count; i++) {
+    m_free_list.push_back(new Entry());
+  }
+}
+
+SimplePolicy::~SimplePolicy() {
+  ldout(cct, 20) << dendl;
+
+  for (auto it: m_cache_map) {
+    Entry* entry = reinterpret_cast<Entry*>(it.second);
+    delete entry;
+  }
+
+  for(auto it : m_free_list) {
+    Entry* entry = it;
+    delete entry;
+  }
+}
+
+cache_status_t SimplePolicy::alloc_entry(std::string file_name) {
+  ldout(cct, 20) << "alloc entry for: " << file_name << dendl;
+
+  m_free_list_lock.lock();
+
+  //TODO(): make the max inflight ops configurable
+  if (m_free_list.size() && (inflight_ops < 128)) {
+    Entry* entry = m_free_list.front();
+    ceph_assert(entry != nullptr);
+    m_free_list.pop_front();
+    m_free_list_lock.unlock();
+
+    {
+      RWLock::WLocker wlocker(m_cache_map_lock);
+      m_cache_map[file_name] = entry;
+    }
+    update_status(file_name, OBJ_CACHE_SKIP);
+    return OBJ_CACHE_NONE;
+  }
+
+  m_free_list_lock.unlock();
+  // if there's no free entry, return skip to read from rados
+  return OBJ_CACHE_SKIP;
+}
+
+cache_status_t SimplePolicy::lookup_object(std::string file_name) {
+  ldout(cct, 20) << "lookup: " << file_name << dendl;
+
+  RWLock::RLocker rlocker(m_cache_map_lock);
+
+  auto entry_it = m_cache_map.find(file_name);
+  // simplely promote on first lookup
+  if (entry_it == m_cache_map.end()) {
+      rlocker.unlock();
+      return alloc_entry(file_name);
```

Comment:
Isn't there is a race here if two requests for the same missing object happen simultaneously?

Both successfully take read `m_cache_map_lock`, find out the object is not in the cache, drop `m_cache_map_lock`. Then in `alloc_entry` one of the threads takes `m_free_list_lock`, gets a free entry, adds it to m_cache_map and returns `OBJ_CACHE_NONE`  (initiating rados read). Then the second thread does the same, overwriting `m_cache_map[file_name]` with its own entry.

It looks like in `alloc_entry` you need to take write `m_cache_map_lock` at the beginning, recheck `m_cache_map[file_name]` and just return `OBJ_CACHE_SKIP` if it already exists. And if it does not, release the lock only after inserting the entry.

Reply:
indeed, it's a bug! Should check the entry status again after get m_free_list_lock. will add a test on this.

Reply:
hi @trociny  currently, the thread model of CacheServer side is very simple. One thread sequentially execute read/write/accept callback of ASIO interface. And different session also share one thread. 
So, based on current simple thread model, i think these code can work well.  And it isn't a bug. 

PS: There just is a race between asio thread and rados callback thread.  

In the future, if we use thread pool at CacheServer side,  your suggestion is very critical. 

thanks your suggestion ~